src/frontend/core/executor/impl/PagerankExecutor.cpp

```diff
@@ -0,0 +1,325 @@
+    processInformation.priority = threadPriority;
+    processInformation.startTimestamp = startTime.count();
+
+    if (!queueTime.empty()) {
+        long sleepTime = atol(queueTime.c_str());
+        processInformation.sleepTime = sleepTime;
+        processData.insert(processInformation);
+        processStatusMutex.unlock();
+        std::this_thread::sleep_for(std::chrono::milliseconds(sleepTime));
+    } else {
+        processData.insert(processInformation);
+        processStatusMutex.unlock();
+    }
+
+    pgrnk_logger.log(
+            "###PAGERANK-EXECUTOR### Started with graph ID : " + graphId + " Master IP : " + masterIP, "info");
+
+    int partitionCount = 0;
+    std::vector<std::future<void>> intermRes;
+    std::vector<std::future<int>> statResponse;
+
+    auto begin = chrono::high_resolution_clock::now();
+
+    std::map<std::string, JasmineGraphServer::workerPartition> graphPartitionedHosts =
+            JasmineGraphServer::getWorkerPartitions(graphId);
+    string partition;
+    string host;
+    int port;
+    int dataPort;
+    std::string workerList;
+
+    std::map<std::string, JasmineGraphServer::workerPartition>::iterator workerIter;
+    for (workerIter = graphPartitionedHosts.begin(); workerIter != graphPartitionedHosts.end(); workerIter++) {
+        JasmineGraphServer::workerPartition workerPartition = workerIter->second;
+        partition = workerPartition.partitionID;
+        host = workerPartition.hostname;
+        port = workerPartition.port;
+        dataPort = workerPartition.dataPort;
+
+        if (host.find('@') != std::string::npos) {
+            host = Utils::split(host, '@')[1];
+        }
+
+        workerList.append(host + ":" + std::to_string(port) + ":" + partition + ",");
+    }
+
+    workerList.pop_back();
+    pgrnk_logger.info("Worker list " + workerList);
+
+    for (workerIter = graphPartitionedHosts.begin(); workerIter != graphPartitionedHosts.end(); workerIter++) {
+        JasmineGraphServer::workerPartition workerPartition = workerIter->second;
+        partition = workerPartition.partitionID;
+        host = workerPartition.hostname;
+        port = workerPartition.port;
+        dataPort = workerPartition.dataPort;
+
+        intermRes.push_back(std::async(
+                std::launch::async, PagerankExecutor::doPageRank, graphId, alpha,
+                iterations, partition, host, port, dataPort, workerList));
+    }
+
+    PerformanceUtil::init();
+
+    std::string query =
+            "SELECT attempt from graph_sla INNER JOIN sla_category where graph_sla.id_sla_category=sla_category.id and "
+            "graph_sla.graph_id='" +
+            graphId + "' and graph_sla.partition_count='" + std::to_string(partitionCount) +
+            "' and sla_category.category='" + Conts::SLA_CATEGORY::LATENCY +
+            "' and sla_category.command='" + PAGE_RANK + "';";
+
+    std::vector<vector<pair<string, string>>> queryResults = perfDB->runSelect(query);
```

Comment:
Is there a chance for us to get empty result here? In that case we will not be able to dereference like `queryResults[0][0]`

Reply:
Yes, that's why I used if condition on line 133

Reply:
Ok